File: problem-381/381.py
import euclidean

solution = 0
primes = [ 2, 3 ]
N = pow( 10, 8 )
for p in xrange( 5, N, 2 ):
    found = True
    for prime in primes:
        if p % prime == 0:
            found = False
            break
        if prime * prime > p:
            break
    if found:
        primes.append( p )
        s = fact = p - 1
        for k in range( p - 1, p - 5, -1 ):
            inv = euclidean.inverse( k, p )
            fact = ( fact * inv ) % p
            s = ( s + fact ) % p
        solution = solution + s
    if p % 100000 == 1:
        print "Progress: %i" % ( 100 * p / N )
print solution

File: problem-381/381.cpp
#include "euclidean.h"
#include "millerrabin.h"
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <utility>
#include <ctime>

typedef unsigned long long int lli;

using namespace std;

inline int S( lli p ) {
    lli s, fact;

    s = fact = p - 1;
    for ( int k = p - 1; k > p - 5; --k ) {
        int inv = inverse( k, p );
        // assert( inv >= 0 );
        fact = ( fact * inv ) % p;
        s = ( s + fact ) % p;
    }
    return s;
}

int main() {
    lli solution = 0;
    int progress = 0;
    lli N = pow( 10, 8 );

    for ( int p = 5; p < N; p += 2 ) {
        if ( isPrime( p ) ) {
            solution += S( p );
        }
        if ( p % 1000000 == 1 ) {
            printf( "Progress: %i%%\n", progress );
            ++progress;
        }
    }
    printf( "%llu\n", solution );
}
